volumes: Implement subpath mount

vvoland opened the pull request:
- Depends on: https://github.com/moby/moby/pull/45784


- Fixes https://github.com/moby/moby/issues/32582
- Supersedes, closes https://github.com/moby/moby/pull/39041



**- What I did**
Make it possible to mount subdirectory of a named volume.

**- How I did it**

`VolumeOptions` now has a `Subpath` field which allows to specify a path relative to the volume that should be mounted as a destination.

Symlinks are supported, but they cannot escape the base volume directory.
A protection against the Time-of-check to Time-of-use is implemented for Linux and Windows to make it not possible to replace the mount source with a symlink that escapes the volume directory.

On Linux, all subpath components are opened with openat, relative to the base volume directory and checked against the volume escape. The final file descriptor is mounted from the /proc/self/fd/<fd> to a temporary mount point owned by the daemon and then passed to the underlying container runtime. Temporary mountpoint is removed after the container is started.

On Windows, all components of the path are locked before the check, and released once the path is already mounted.


**- How to verify it**
`TestRunMountVolumeSubdir` integration test

**- Description for the changelog**
Add `Subpath` field to the `VolumeOptions` making it possible to mount a subpath of a volume.


**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://github.com/moby/moby/assets/5046555/45213ab8-cca6-44c9-b8d0-33d9de8eba0e)

vvoland committed:
container: Change comment into debug log

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
daemon/populateVolumes: Support volume subpath

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
volume/mounts: Rename errors in defer block

To make it easier to distinguish if an output variable is modified.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
volumes: Implement subpath mount

`VolumeOptions` now has a `Subpath` field which allows to specify a path
relative to the volume that should be mounted as a destination.

Symlinks are supported, but they cannot escape the base volume
directory.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
internal/safepath: Add linux implementation

All subpath components are opened with openat, relative to the base
volume directory and checked against the volume escape.
The final file descriptor is mounted from the /proc/self/fd/<fd> to a
temporary mount point owned by the daemon and then passed to the
underlying container runtime.
Temporary mountpoint is removed after the container is started.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
internal/safepath: Add windows implementation

All components of the path are locked before the check, and
released once the path is already mounted.
This makes it impossible to replace the mounted directory until it's
actually mounted in the container.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

thaJeztah commented on contrib/busybox/Dockerfile:
```diff
@@ -10,16 +10,16 @@
 ARG BUSYBOX_VERSION
 ARG BUSYBOX_SHA256SUM
-ADD https://github.com/moby/busybox/releases/download/${BUSYBOX_VERSION}/busybox-w32-${BUSYBOX_VERSION}.exe /bin/busybox.exe
+ADD https://frippery.org/files/busybox/busybox-w32-${BUSYBOX_VERSION}.exe /bin/busybox.exe
```
Looks like we need to update https://github.com/moby/busybox

Not sure how we got the binaries there (perhaps that was done manually? maybe we should have a github action in that repo /cc @crazy-max )

cpuguy83 commented on volume/mounts/mounts.go:
```diff
@@ -95,50 +98,93 @@ func (m *MountPoint) Cleanup() error {
+			path = newPath
+			logrus.Debugf("mounting (%s|%s) via %s", path, subpath, newPath)
+
+			// TODO: Find a way to have it only for integration tests
+			if strings.HasSuffix(subpath, "iwanttobehackedwithtoctou") {
+				prev, prevErr := filepath.EvalSymlinks(combinedPath)
+				logrus.WithError(prevErr).Debugf("%s resolves to %s", combinedPath, prev)
+
+				for i := 0; i < 10; i++ {
+					time.Sleep(time.Second * time.Duration(i))
+					now, err := filepath.EvalSymlinks(combinedPath)
+					logrus.WithError(err).WithField("retry", i).Debugf("%s resolves to %s", combinedPath, now)
+
+					if now != prev || !errors.Is(err, prevErr) {
+						if !os.IsNotExist(err) {
+							break
+						}
+					}
+				}
+			}
```
Maybe turn it into a unit test and set some unexported field on `*Mountpoint`?

cpuguy83 commented on pkg/safepath/join_unsafe.go:
```diff
@@ -0,0 +1,31 @@
+)
+
+// This is unsafe and suspectible to Time-Of-Check Time-Of-Use attack.
+func Join(path, subpath string) (string, func() error, error) {
```
I think I'd prefer to error out on systems that we don't have an implementation of a safe join.

vvoland replied:
Done

vvoland commented on pkg/safepath/join_test.go:
```diff
@@ -0,0 +1,107 @@
+	assert.Assert(t, os.WriteFile(filepath.Join(dir, "foo"), []byte("bar"), 0o744), "failed to create file 'foo'")
+	assert.Assert(t, os.Mkdir(filepath.Join(dir, "subdir"), 0o744), "failed to create directory 'subdir'")
+
+	//assert.Assert(t, os.Symlink(filepath.Join(dir, "subdir"), filepath.Join(dir, "subdir_link_absolute")), "failed to create absolute symlink to directory 'subdir'")
```
Turns out there's no way for `Openat2` to resolve absolute links (even if they don't escape the base directory).

We can either:
1. Don't promise that absolute (or even any) links will work as a mount source.
2. Drop the `openat2` and only use the "fallback" `Openat` implementation.

Personally, I'm in favor of 1 - no promise that absolute links work.
WDYT?

vvoland replied:
Went with 3 - resolve the symlinks first and then traverse path without symlinks.

vvoland replied:
I didn't really consider copying the Kubernetes implementation. 
It seems like we could be easily just use the Kubernetes [doSafeOpen](https://github.com/kubernetes/kubernetes/blob/55fb1805a1217b91b36fa8fe8f2bf3a28af2454d/pkg/volume/util/subpath/subpath_linux.go#L526-L609) as the fallback implementation. Using the whole implementation would require some major modifications (but you probably didn't mean to copy their whole `subpath` package.

I'm okay with it If everybody is fine with it and there are no legal objections (k8s is Apache 2, so probably not, but IANAL).

thaJeztah commented on api/server/router/container/container_routes.go:
```diff
@@ -597,6 +597,11 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 					return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
 				}
 			}
+			if m.VolumeOptions != nil {
+				if m.VolumeOptions.SubPath != "" {
+					return errdefs.InvalidParameter(errors.New("VolumeOptions.SubPath needs API v1.44 or newer"))
+				}
+			}
```
I wonder if this one should be an error or just be reset on older API versions. I know we reset such things for most (I guess the recursive-readonly was an exception because it was from a security perspective, and ignoring it would weaken security)

laurazard replied:
+1 for error here, failing silently would be confusing and I'd be cursing at the computer failing to realise why I can't find my files where I expect them to be.

corhere commented on daemon/volumes_unix.go:
```diff
@@ -12,19 +12,37 @@ import (
+func (daemon *Daemon) setupMounts(c *container.Container) ([]container.Mount, func() error, error) {
+	var cleanups []func() error
+	cleanup := func() error {
+		var allErr error
+		for _, c := range cleanups {
+			err := c()
+			if err != nil {
+				if allErr == nil {
+					allErr = err
+				} else {
+					allErr = multierror.Append(allErr, err)
+				}
+			}
+		}
+
+		return allErr
```
Nit: I don't think it's worth the effort to avoid returning a multierror wrapping one error.
```suggestion
		var errs *multierror.Error
		for _, c := range cleanups {
			errs = multierror.Append(errs, c())
		}
		return errs.ErrorOrNil()
```

corhere reviewed (changes requested):
Safely traversing the filesystem is really tricky to get right

vvoland reviewed (commented):
Thank you for the thorough review!

I addressed the easier comments, I will continue working on the rest next week.

corhere commented on pkg/safepath/join_linux.go:
```diff
@@ -0,0 +1,204 @@
+// getFdRealpath returns the real and procfs path for the given file descriptor.
+func getFdRealpath(fd int) (realPath string, procFsPath string, outErr error) {
+	procFd := procFsFdPath(fd)
+	realPath, err := os.Readlink(procFd)
```
The `/proc/self/fd/[n]` link could be a link to a symlink if a symlink was opened with `open(O_PATH|O_NOFOLLOW)`. Since the function takes an arbitrary file descriptor, it is not safe in general to assume that readlink would produce the same result as realpath. Only the caller can know if that optimization is safe to use.

vvoland committed:
daemon/oci: Extract side effects from withMounts

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

corhere reviewed (changes requested):
Remember that [raw string literals](https://go.dev/ref/spec#:~:text=Raw%20string%20literals%20are,string%20may%20contain%20newlines.) exist as they would make the Windows path string literals much more readable in the tests.

vvoland committed:
internal/safepath: Adapt k8s openat2 fallback

Adapts the function source code to the Moby codebase.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
Introduce a helper that collects cleanup functions

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

corhere reviewed (commented):
Partial review; I will likely pick back up tomorrow

vvoland commented on daemon/containerfs_linux.go:
```diff
@@ -62,17 +62,18 @@ func (daemon *Daemon) openContainerFS(container *container.Container) (_ *contai
-	if err != nil {
-		return nil, err
-	}
 	defer func() {
 		if err != nil {
 			_ = container.UnmountVolumes(daemon.LogVolumeEvent)
 		}
 	}()
 
+	mounts, cleanup, err := daemon.setupMounts(container)
+	if err != nil {
+		return nil, err
+	}
+	defer cleanup()
```
It isn't necessary if `setupMounts` fails (will be a no-op). If it succeeds then we always cleanup the temporary mounts/file locks first (call the cleanup from `setupMounts`), before the `MountPoint.Cleanup` so we keep the promise of caller calling the `cleanup` received from `daemon.setupMounts`/`MountPoint.Setup`.

corhere reviewed (commented):
I've reviewed up to (but not including) internal/safepath/join_linux.go

corhere commented on volume/mounts/validate_test.go:
```diff
@@ -24,6 +24,9 @@ func TestValidateMount(t *testing.T) {
 		{mount.Mount{Type: mount.TypeVolume}, errMissingField("Target")},
 		{mount.Mount{Type: mount.TypeVolume, Target: testDestinationPath, Source: "hello"}, nil},
 		{mount.Mount{Type: mount.TypeVolume, Target: testDestinationPath}, nil},
+		{mount.Mount{Type: mount.TypeVolume, Target: testDestinationPath, Source: "hello", VolumeOptions: &mount.VolumeOptions{Subpath: "world"}}, nil},
+		{mount.Mount{Type: mount.TypeVolume, Target: testDestinationPath, Source: "hello", VolumeOptions: &mount.VolumeOptions{Subpath: "world"}}, nil},
```
Why is this test case run twice?

vvoland commented on daemon/oci_linux.go:
```diff
@@ -508,7 +508,7 @@ func inSlice(slice []string, s string) bool {
-func withMounts(daemon *Daemon, daemonCfg *configStore, c *container.Container) coci.SpecOpts {
+func withMounts(daemon *Daemon, daemonCfg *configStore, c *container.Container, ms []container.Mount) coci.SpecOpts {
 	return func(ctx context.Context, _ coci.Client, _ *containers.Container, s *coci.Spec) (err error) {
 		if err := daemon.setupContainerMountsRoot(c); err != nil {
```
Hmm, I think it's probably good to have `daemon.setupContainerMountsRoot(c)` first.
I extracted the side effects into a separate function (and also did the same for Windows).

corhere commented on integration/internal/container/container.go:
```diff
@@ -71,3 +73,29 @@ func Run(ctx context.Context, t *testing.T, client client.APIClient, ops ...func
+}
+
+// WaitToEnd waits for the container to end running and returns its output.
+func WaitToEnd(ctx context.Context, client client.APIClient, id string) (ContainerOutput, error) {
```
Good call. I'd be happy with either of your proposed names.

corhere commented on integration/volume/mount_test.go:
```diff
@@ -0,0 +1,186 @@
+
+	opts := []func(*container.TestContainerConfig){
+		container.WithMount(mount),
+		container.WithCmd("sh", "-c", initCmd+"; ls -lah /volume/hack/"),
```
No reason; I just didn't realize that was a deliberate design choice.

thaJeztah reviewed (commented):
Started to review this; let me post what I had so far (but will continue looking at some point 😅 )

vvoland commented on internal/safepath/errors.go:
```diff
@@ -0,0 +1,9 @@
+	"github.com/pkg/errors"
+)
+
+var ErrorNotAccessible = errdefs.NotFound(errors.New("path is not accessible"))
+var ErrorEscapesBase = errdefs.InvalidParameter(errors.New("joined path escapes the base directory"))
```
Ah, thanks, changed.
Wrote a small Godoc for these, although I must admit it's not something that can't be inferred from the name and message alone 😄

vvoland commented on daemon/volumes_windows.go:
```diff
@@ -1,33 +1,57 @@
+		var errs *multierror.Error
+		for _, c := range cleanups {
+			errs = multierror.Append(errs, c())
+		}
```
Oh, TIL it's finally in the stdlib. 😄 
Changed to stdlib errors, thanks!

thaJeztah replied:
Yes, I honestly missed it as well, but there was a colleague complaining that they didn't like the output format (newline separated). I guess that's fine for now (we can still do something with the output if we really need to)

vvoland commented on volume/mounts/mounts.go:
```diff
@@ -98,26 +121,42 @@ func (m *MountPoint) Cleanup() error {
+// frees up the resources that were needed to guarantee that the path
+// still points to the same target (to avoid TOCTOU attack).
+//
+// Cleanup function doesn't need to be called when error is returned.
```
A noop cleanup still needs to be returned in some cases when `err == nil` anyway, so I don't think it makes much sense to also make it possible to be nil.

vvoland commented on volume/mounts/mounts.go:
```diff
@@ -108,20 +108,16 @@ func (m *MountPoint) Setup(mountLabel string, rootIDs idtools.Identity, checkFun
-		var sourcePath string
-		sourcePath, err = filepath.EvalSymlinks(m.Source)
-		if err != nil {
+		sourcePath, evalErr := filepath.EvalSymlinks(m.Source)
```
This is what I had originally, but I changed it because the return variable names are exposed in the Godoc. See: https://github.com/moby/moby/pull/45687#discussion_r1245632003

cpuguy83 commented on volume/mounts/mounts.go:
```diff
@@ -95,30 +99,27 @@ func (m *MountPoint) Cleanup() error {
 // The, optional, checkFun parameter allows doing additional checking
 // before creating the source directory on the host.
-func (m *MountPoint) Setup(mountLabel string, rootIDs idtools.Identity, checkFun func(m *MountPoint) error) (path string, err error) {
+func (m *MountPoint) Setup(mountLabel string, rootIDs idtools.Identity, checkFun func(m *MountPoint) error) (outPath string, cleanup func() error, outErr error) {
```
Not sure I understand this comment.
Why does this matter?

cpuguy83 commented on internal/safepath/safepath.go:
```diff
@@ -36,15 +41,20 @@ func (s *SafePath) Close() error {
 func (s *SafePath) Path() string {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 	return s.path
```
What happens if something tries to use this `SafePath` after close (because s.path is zeroed out currently).

vvoland replied:
Theoretically:
1. If you use the result of `.Path()` obtained before path was `.Close()`d, that path is not guaranteed to point to the same thing as before `.Close()` was called or even exist at all. 
2. If you call `.Path()` after `.Close()` you get an empty path, and passing it anywhere should error with NotFound error. I guess we could add an explicit panic in this case?

cpuguy83 replied:
Panic seems like the thing to do here.

vvoland replied:
Added a panic.

cpuguy83 reviewed (changes requested):
2b8c1cac0ca8ef2aca347a832ffde4406d8be98d seems incorrectly titled and doesn't really give any detail as to why (nor is the new function/type consumed in that commit).
Maybe it should be merged with the commit where `TestRunMountVolumeSubddir` is added?

vvoland commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,200 @@
+// error was returned.
+func safeOpenFd(path, subpath string) (int, error) {
+	// Open base volume path (_data directory).
+	prevFd, err := unix.Open(path, unix.O_PATH|unix.O_DIRECTORY|unix.O_CLOEXEC|unix.O_NOFOLLOW, 0)
```
Done, I wondered if there's already a package that just wraps the `x/sys/unix` with the EINTR retry, but the only package I found is this: https://github.com/AkihiroSuda/x-sys-unix-auto-eintr (hmmm, that name sounds very familiar! 😄), but unfortunately it's no longer maintained. So I just created a `internal/unix_noeintr`. Wondering if it would be better suited in the `moby/sys` or some other package though.

cpuguy83 commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,145 @@
+}
+
+func procFsFdPath(fd int) string {
+	return fmt.Sprintf("/proc/self/fd/%d", fd)
```
```suggestion
	return "/proc/self/fd/"+strconv.Itoa(fd)
```
This save an allocation and is about 2x as fast.

cpuguy83 replied:
Also it looks like this is only used in one place.
I think we can inline this.

vvoland replied:
Done, thanks!

cpuguy83 commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,145 @@
+	isDir := (stat.Mode & unix.S_IFMT) == unix.S_IFDIR
+	if isDir {
+		return os.MkdirTemp(base, "safe-mount")
+	} else {
```
No need for `else`

vvoland replied:
Done, thanks.

cpuguy83 commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,145 @@
+		log.G(context.TODO()).WithField("path", path).Debug("removing safe temp mount")
+
+		if err := mount.Unmount(path); err != nil {
+			if errors.Is(err, os.ErrNotExist) {
```
There is no such case here.
`unmount` would return EINVAL if there was no mountpoint (not an ErrNotExist) but `mount.Unmount` actually ignores the error and returns nil.

vvoland replied:
Thanks, good catch.
Replaced with a raw unix syscall with EINTR retry and handled EINVAL.

cpuguy83 commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,145 @@
+			return errors.Wrapf(err, "error unmounting safe mount %s", path)
+		}
+		if err := os.Remove(path); err != nil {
+			return errors.Wrapf(err, "failed to delete temporary safe mount")
```
```suggestion
			if errors.Is(err, os.ErrNotExist) {
				return nil
			}
			return errors.Wrapf(err, "failed to delete temporary safe mount")
```

vvoland replied:
Done, added a warning in this case though, as I think it's not expected for it to not exist at this point.

cpuguy83 commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,145 @@
+			return errors.Wrapf(err, "failed to delete temporary safe mount")
+		}
+
+		path = ""
```
This looks redundant.

vvoland replied:
Removed, thanks!

cpuguy83 commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,145 @@
+
+	prevFd, err := unix_noeintr.Open(path, unix.O_PATH|unix.O_DIRECTORY|unix.O_CLOEXEC|unix.O_NOFOLLOW, 0)
+	if err != nil {
+		return -1, ErrNotAccessible
```
Seems like this could lead to misunderstood problems since we are masking the actual error.
Maybe use the new multi-error support?

vvoland replied:
I think I'll just turn it into a rich error struct at this point that will contain the original error and the specific path information.

cpuguy83 commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,145 @@
+// After the returned path is consumed by the caller, it is his responsibility
+// to call Close on the returned SafePath object which will unmount the
+// temporary file/directory and remove it.
+func Join(path, subpath string) (*SafePath, error) {
```
We are doing logging in this package, can we pass down a context?

cpuguy83 commented on internal/safepath/join_unsafe.go:
```diff
@@ -0,0 +1,35 @@
+)
+
+// This is unsafe and suspectible to Time-Of-Check Time-Of-Use attack.
+func Join(path, subpath string) (*SafePath, error) {
```
I think this should just error out.

vvoland replied:
The unsafe implementation is removed in latter commit.

cpuguy83 commented on volume/mounts/mounts.go:
```diff
@@ -75,14 +76,36 @@ type MountPoint struct {
+		}
+
+		base, sub := p.SourcePath()
+		log.G(context.TODO()).WithFields(logrus.Fields{
```
If we log here, we should accept a context in the function.

cpuguy83 commented on volume/mounts/mounts.go:
```diff
@@ -75,14 +76,36 @@ type MountPoint struct {
+		}).Warn("cleaning up SafePath that hasn't been cleaned up by the caller")
+
+		if err := p.Close(); err != nil {
+			return errors.Wrapf(err, "error cleaning up safepath: %v", p)
```
This error condition is intriguing.
I'm not sure what to do here.
What do we expect the user to do in this situation? Seems like this would block the container from moving to a shutdown state.

vvoland replied:
Hmm, maybe we can ignore it at this point and only log a warning. If the cleanup failure makes it impossible to unmount the base volume, the `m.Volume.Unmount` will fail anyway. And if that still succeeds, then perhaps the failure wasn't critical.

Also, if this has been reached, this means that the caller of `Setup` didn't call the cleanup on his own, which probably means that something went really wrong in some other place.

cpuguy83 reviewed (changes requested):
Another round here.

One thing I'm not liking is the interaction with `Mountpoint` which is already complex before the change.

I'm wondering if we could treat subpath mounts as a volume driver that wraps other volume drivers.
Would this simplify the setup and cleanup handling?

I'm thinking most calls to this wrapper driver would just be passthroughs to the underlying driver but then `Mount` and `Unmount` perform additional steps to deal with the subpath.

vvoland commented:
> I'm wondering if we could treat subpath mounts as a volume driver that wraps other volume drivers.
Would this simplify the setup and cleanup handling?

> I'm thinking most calls to this wrapper driver would just be passthroughs to the underlying driver but then Mount and Unmount perform additional steps to deal with the subpath.


The safepath cleanup is done separately from the `MountPoint.Cleanup` (well, it's done here as well, but only as a last resort if for some reason caller didn't do it). The reason for that is to remove the temporary mount (on Linux) and unlock the path components (on Windows) as soon as the container is started and not hold it for the entire lifetime of the container.

Also, wrapping it as volume driver is another layer of indirection which I think is more confusing to follow than just making the `MountPoint` a little bit more complex. Especially since this is expected to also work with other drivers.

vvoland committed:
volumes/subpath: Plumb context

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented on internal/cleanups/composite.go:
```diff
@@ -1,40 +1,74 @@
+func call(ctx context.Context, cleanups []func(ctx context.Context) error) error {
 	var errs error
+	// Disallow cancellation, but pass all values.
+	ctx = notCancellable{parent: ctx}
```
Eh, I don't like this, and would prefer to just not pass a context here, but I don't think we want a context to be cancelled here, and we only need the values of the context for the `log.G(ctx)` to work as expected.

vvoland replied:
Hmm, looks like it's in 1.21: https://pkg.go.dev/context@master#WithoutCancel

vvoland reviewed (commented):
So, I plumbed the context through, but may have opened a can of worms.

thaJeztah commented:
@laurazard are you able to help review? 🤗

vvoland commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,151 @@
+
+func procFsFdPath(fd int) string {
+	pid := unix.Getpid()
+	return fmt.Sprintf("/proc/%d/fd/%d", pid, fd)
```
Actually, I had to revert this one - mounting `/proc/self/` fails with `EACCES` when running with `Enhanced Container Isolation` in DD (which uses the `sysbox` runtime underneath).

I reached out to the relevant team to discuss this, but for now looks like we should just use an explicit pid path.

laurazard commented on internal/safepath/common_test.go:
```diff
@@ -0,0 +1,31 @@
+		{name: "absolute", subpath: "/etc/passwd", result: false},
+		{name: "backtrack", subpath: "/volume/../", result: false},
+		{name: "backtrack inside", subpath: "/volume/sub/../", result: true},
+		{name: "relative path", subpath: "./rel", result: false},
```
> Both path and subpath are expected to be absolute paths

should we make this explicit in the godoc?

laurazard commented on volume/mounts/mounts.go:
```diff
@@ -188,24 +184,6 @@ func (m *MountPoint) Setup(mountLabel string, rootIDs idtools.Identity, checkFun
 			volumePath = safePath.Path()
-
-			// TODO: Find a way to have it only for integration tests
-			if strings.HasSuffix(subpath, "iwanttobehackedwithtoctou") {
```
It would be nice to keep the TOCTOU test – I wonder if we can find a way to keep it without touching the implementation.

vvoland replied:
Yeah, I have no idea how to do it nicely with the current way integration tests are run.
We still have the `TestJoinWithSymlinkReplace` unit test that checks it though.

laurazard commented on internal/safepath/errors.go:
```diff
@@ -0,0 +1,42 @@
+package safepath
+
+// ErrNotAccessible is returned by Join when the resulting path doesn't exist,
+// it's not accessible or any of the path component was replaced with a symlink
```
```suggestion
// is not accessible, or any of the path components was replaced with a symlink
```

vvoland replied:
Thanks, changed!

laurazard commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,148 @@
+// resolve to a path outside of path and returns a path to a temporary file that is
+// a bind mount to the the exact same file/directory that was validated.
+//
+// After the returned path is consumed by the caller, it is his responsibility
+// to call Close on the returned SafePath object which will unmount the
+// temporary file/directory and remove it.
```
```suggestion
// After use, it is the caller's responsibility to call Close on the returned
// SafePath object, which will unmount the temporary file/directory
// and remove it.
```

vvoland replied:
Thanks, changed!

cpuguy83 commented on daemon/start.go:
```diff
@@ -152,7 +152,19 @@ func (daemon *Daemon) containerStart(ctx context.Context, daemonCfg *configStore
+
+	if m, cleanup, err := daemon.setupMounts(ctx, container); err != nil {
+		return err
+	} else {
```
Uneccessary `else`

vvoland replied:
It's actually needed, because `m` and `cleanup` are scoped to the if.

corhere replied:
[“Programs must be written for people to read, and only incidentally for machines to execute.”
― Harold Abelson, Structure and Interpretation of Computer Programs](https://www.goodreads.com/quotes/9168-programs-must-be-written-for-people-to-read-and-only)

The scoping of `m` is only a problem for the person writing the code. [Indenting error flow](https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow) improves the readability of the code. Readability of the code takes priority.

vvoland replied:
Scope can also provide useful information to the reader.
In my view the `else` doesn't decrease readability here and the scope is also a useful information to the reader.
I agree that it's much more readable to indent the error flow and return as early as possible, instead of indenting the whole happy path covering majority of the function. In this case only 2 lines are indented, which are tightly coupled to the result of `setupMounts`.

But, if you're the third reader to agree with this suggestion, then I must accept that it's more readable and my subjective perception of readability is less likely to be shared by an average reader, so I'll just conform 😄.

Changed.

cpuguy83 reviewed (approved):
I have reviewed this several times now.
It seems like we are doing all the right things here to prevent subpath's from traversing outside the volume root.

LGTM

cpuguy83 commented:
Looks like a rebase is needed

> docs/api/version-history.md

vvoland commented:
Thank you for the review!
Rebased.

nathancahill commented:
Ah this is fantastic. Hoping for a quick merge. Nice work!

corhere commented on internal/unix_noeintr/fs_unix.go:
```diff
@@ -0,0 +1,82 @@
+// Wrappers for unix syscalls that retry on EINTR
+
+import (
+	"github.com/pkg/errors"
```
```suggestion
	"errors"
```

vvoland replied:
Ah, autoimport.. Fixed!

corhere reviewed (commented):
I'm submitting a pending review that's a couple months old. Several comments may be outdated.

vvoland commented on internal/safepath/common.go:
The intent was to keep only the code related to the `SafePath` struct inside `safepath.go` and have helper function used by other files (`join_linux.go`, `join_windows.go` and `k8s_safeopen_linux.go`)  in `common.go`.
Having them in `common.go` instead of `join.go` allows to separate the `isLocalTo` unit tests (`common_test.go`) from the `Join` unit tests (`join_test.go`).

vvoland commented on integration/volume/mount_test.go:
```diff
@@ -0,0 +1,188 @@
+
+			ctrName := strings.ReplaceAll(t.Name(), "/", "_")
+			create, creatErr := client.ContainerCreate(ctx, &cfg, &hostCfg, &network.NetworkingConfig{}, nil, ctrName)
+			if tc.createErr != "" {
+				assert.ErrorContains(t, creatErr, tc.createErr)
+				return
+			} else {
+				assert.NilError(t, creatErr, "container creation failed")
+			}
+			id := create.ID
+
+			startErr := client.ContainerStart(ctx, id, types.ContainerStartOptions{})
+			if tc.startErr != "" {
+				assert.ErrorContains(t, startErr, tc.startErr)
+				return
+			} else {
+				assert.NilError(t, startErr)
+			}
+
+			defer client.ContainerRemove(ctx, id, types.ContainerRemoveOptions{Force: true})
```
Good catch, thanks, done.

vvoland commented on internal/safepath/common.go:
```diff
@@ -0,0 +1,59 @@
+// - is not empty
+// - on Windows, is not a reserved name such as "NUL"
+//
+// If isRelativeTo(base, path) returns true, then it's there's always a `subpath` for which the following conditions are true:
+// - filepath.IsLocal(subpath)
+// - filepath.Join(base, subpath) == path
```
Changed to your suggestion.

vvoland commented on internal/safepath/common.go:
```diff
@@ -0,0 +1,59 @@
+// - filepath.Join(base, subpath) == path
+//
+// isRelativeTo is a purely lexical operation. In particular, it does not account for the effect of any symbolic links that may exist in the filesystem.
+func isRelativeTo(path, subpath string) bool {
+	rel, err := filepath.Rel(path, subpath)
+	if err != nil {
+		return false
+	}
+
+	return filepath.IsLocal(rel)
+}
```
Done.

vvoland commented on internal/safepath/safepath.go:
```diff
@@ -0,0 +1,60 @@
+
+// SourcePath returns the source path the safepath points to.
+func (s *SafePath) SourcePath() (string, string) {
+	// No mutex lock because these are not modified.
```
Done.

vvoland commented on internal/safepath/safepath.go:
```diff
@@ -0,0 +1,60 @@
+	path                      string
+	sourceBase, sourceSubpath string
+	cleanup                   func() error
+	mutex                     sync.Mutex
```
Done.

vvoland commented on daemon/oci_linux.go:
```diff
@@ -507,48 +507,49 @@ func inSlice(slice []string, s string) bool {
+	defer func() {
 		if err != nil {
-			return err
+			daemon.cleanupSecretDir(c)
```
Done.

vvoland commented on daemon/oci_linux.go:
```diff
@@ -507,48 +507,49 @@ func inSlice(slice []string, s string) bool {
-			}
-		}()
+// setupContainerDirs sets up base container directories (root, ipc, tmpfs and secrets).
+func (daemon *Daemon) setupContainerDirs(c *container.Container) (_ []container.Mount, err error) {
```
Moved to container_operations_unix.go

vvoland committed:
internal/safepath: Handle EINTR in unix syscalls

Handle EINTR by retrying the syscall.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
internal/safepath: Import k8s safeopen function

For use as a soft fallback if Openat2 is not available.
Source: https://github.com/kubernetes/kubernetes/blob/55fb1805a1217b91b36fa8fe8f2bf3a28af2454d/pkg/volume/util/subpath/subpath_linux.go

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

neersighted commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,146 @@
+	pid := strconv.Itoa(unix.Getpid())
+	// Using explicit pid path, because /proc/self/fd/<fd> fails with EACCES
+	// when running under "Enhanced Container Isolation" in Docker Desktop.
+	mountSource := "/proc/" + pid + "/fd/" + strconv.Itoa(fd)
```
(my specific objection to this is that a fake procfs is not a real target to develop against, sysbox needs to be compatible with the kernel)

thaJeztah commented:
@vvoland needs a rebase 😢 (sorry!)

thaJeztah reviewed (commented):
I went through all the commits, and looks great! (thanks for cleaning them up! (there's even some in them that _could_ already be merged separate, but they're so small that it's probably fine to keep them as part of this))

Most of my comments are non-blocking (couple of nits; some of those "nice to do" while where adding the new code),

The ones to look at;

- The exported errors; not critical, but if we don't need them exported, it'd be good to un-export them (then we don't have to worry about anything else consuming them)
- The (potential) duplicate paths: I think it's fine to look at that in a follow-up, but I _think_ this is something that should be fixed (or at least; I can expect there could be situations where having duplicates could cause issues, but that's just "gut feeling"). I don't want to see that as a blocker (just "don't do that" would suffice for now, but let's make sure we catch those cases).

corhere commented on volume/mounts/linux_parser.go:
```diff
@@ -94,8 +94,20 @@ func (p *linuxParser) validateMountConfigImpl(mnt *mount.Mount, validateBindSour
+				return &errMountConfig{mnt, ErrAnonymousVolumeWithSubpath}
+			}
+
+			path := mnt.VolumeOptions.Subpath
+
+			if !filepath.IsLocal(path) {
```
> (which we should probably replace with `filepath` but that's another topic).

That replacement would change the semantics when built for Windows, so I would advise against it so long as `mounts.NewLinuxParser()` is available on Windows.

corhere commented on volume/mounts/mounts.go:
```diff
@@ -127,16 +164,34 @@ func (m *MountPoint) Setup(mountLabel string, rootIDs idtools.Identity, checkFun
+				}
+				return "", noCleanup, err
+			}
+			m.safePaths = append(m.safePaths, safePath)
```
Why? What's wrong with letting users mount a volume to more than one mountpoint if they want to?

It may prove very tricky to get the check right on Windows (unless we use `os.SameFile` or similar) [because the Windows filesystem is not always case insensitive.](https://devblogs.microsoft.com/commandline/per-directory-case-sensitivity-and-wsl/)

vvoland replied:
I don't think we need to restrict it.
It's safe for multiple `SafePath` to be created from the same `[base, subpath]` path.

On Linux each subpath gets its own independent temporary file/directory the target path gets bind mounted at so cleaning up one `SafePath` doesn't affect others.
On Windows, that will just be another set of open handles locking each part of the target path.

thaJeztah replied:
Maybe I had my wires crossed and was thinking this was for the _destination_ (I'd have to look at the code again), if that's not at hand here, then ignore my comment.

My main concern here was that ending up with duplicate mounts can be troublesome, and we for sure had cases where those cases weren't detected properly (some due to matching happening case-sensitive); such cases have led to users scratching their heads why the "volume was empty" (because it was shadowed by another mount) etc. For such cases (where that being intentional would be "rare") I'd prefer to produce an error instead of silently accepting it (errors can be removed, but the revers (accepting now, and making it an error later) is much harder).

We do have some detection in place for those;

```bash
docker create -v hello:/foo -v world:/foo alpine
Error response from daemon: Duplicate mount point: /foo
```

Although not sure if our detection currently is complete (but possibly we're de-duplicating this case);

```bash
docker create -v /foo -v /foo alpine
552b828c29f596bdb0a0380de4ac1cd48bc9e93a6931540a437be618e82b399a
```

corhere reviewed (commented):
The meat of the implementation looks good to me! I'm very particular about error handling and `cleanups.Composite`, however.

I have a small request to make: please rearrange the commits a bit to keep the amount of code churn to a minimum when viewing the commits individually. In particular:
- squash a81a942834b281b9634ccf5970b4a33cf3a4c699 into a5eb326e1bbad99ac8de0bacf06a13157afdf20b
- add the `cleanups.Composite` implementation (6dba3c412e3e28b8f88c7a218338601c37351a42) before 934ab3378e1d13efa7a6ff7a8623ad46bea2625a instead of deduplicating the cleanup handling after.

I recognize it's not a trivial ask, so [I've already done the work for you.](https://github.com/moby/moby/compare/master...corhere:moby:volume-mount-subpath?expand=1)

vvoland commented on internal/safepath/join_linux.go:
```diff
@@ -0,0 +1,150 @@
+		return nil, err
+	}
+
+	fd, err := safeOpenFd(base, subpart)
+	if err != nil {
+		return nil, err
+	}
+
+	defer unix_noeintr.Close(fd)
+
+	tmpMount, err := tempMountPoint(fd)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to create temporary file for safe mount")
+	}
+
+	pid := strconv.Itoa(unix.Getpid())
```
Thanks, done!

vvoland commented on internal/safepath/join_windows.go:
```diff
@@ -0,0 +1,85 @@
+			}
+			return nil, errors.Wrapf(err, "failed to lock file %s", fullPath)
+		}
+		cleanups.Add(func(context.Context) error {
+			return windows.CloseHandle(handle)
+		})
```
Good point, thanks. Also done the same to the windows calls in `lockFile`.

vvoland commented on volume/mounts/parser.go:
```diff
@@ -11,6 +11,14 @@ import (
 // It's used by both LCOW and Linux parsers.
 var ErrVolumeTargetIsRoot = errors.New("invalid specification: destination can't be '/'")
 
+// ErrAnonymousVolumeWithSubpath is returned when Subpath is specified for
+// anonymous volume.
+var ErrAnonymousVolumeWithSubpath = errors.New("must not set Subpath when using anonymous volumes")
```
Unexported. I didn't unexport the `ErrVolumeTargetIsRoot` which was there before this PR though.

vvoland commented on internal/cleanups/composite.go:
```diff
@@ -1,40 +1,70 @@
-func call(cleanups []func() error) error {
+func call(ctx context.Context, cleanups []func(ctx context.Context) error) error {
 	var errs error
+	// Disallow cancellation, but pass all values.
```
Done.

vvoland commented on daemon/containerfs_linux.go:
```diff
@@ -63,14 +63,14 @@ func (daemon *Daemon) openContainerFS(container *container.Container) (_ *contai
+		cleanup(context.TODO())
 		if err != nil {
-			_ = container.UnmountVolumes(daemon.LogVolumeEvent)
+			_ = container.UnmountVolumes(context.TODO(), daemon.LogVolumeEvent)
```
Done

vvoland commented on daemon/volumes_windows.go:
```diff
@@ -1,33 +1,57 @@
 // BUGBUG TODO Windows containerd. This would be much better if it returned
 // an array of runtime spec mounts, not container mounts. Then no need to
 // do multiple transitions.
+func (daemon *Daemon) setupMounts(c *container.Container) (_ []container.Mount, _ func() error, outErr error) {
```
Fixed, thanks!

corhere commented on daemon/create_unix.go:
```diff
@@ -85,10 +89,41 @@ func (daemon *Daemon) populateVolumes(c *container.Container) error {
+		return err
+	}
+
+	volumePath, cleanup, err := mnt.Setup(ctx, c.MountLabel, daemon.idMapping.RootPair(), func(m *volumemounts.MountPoint) error { return nil })
```
The `checkFun` parameter is optional. No need to pass a no-op function.
https://github.com/moby/moby/blob/b1c112d35e4f1921ca82c9f76d86e5c92c2a9280/volume/mounts/mounts.go#L150-L154
```suggestion
	volumePath, cleanup, err := mnt.Setup(ctx, c.MountLabel, daemon.idMapping.RootPair(), nil)
```

vvoland replied:
Oh, thanks, I forgot about that!

corhere commented on daemon/start.go:
```diff
@@ -138,7 +139,7 @@ func (daemon *Daemon) containerStart(ctx context.Context, daemonCfg *configStore
 			container.Reset(false)
 
-			daemon.Cleanup(container)
+			daemon.Cleanup(ctx, container)
```
```suggestion
			daemon.Cleanup(compatcontext.WithoutCancel(ctx), container)
```

corhere commented on volume/mounts/windows_parser.go:
```diff
@@ -258,7 +259,19 @@ func (p *windowsParser) validateMountConfigReg(mnt *mount.Mount, additionalValid
+		anonymousVolume := len(mnt.Source) == 0
+		if mnt.VolumeOptions != nil && mnt.VolumeOptions.Subpath != "" {
+			if anonymousVolume {
+				return &errMountConfig{mnt, fmt.Errorf("must not set Path when using anonymous volumes")}
```
```suggestion
				return &errMountConfig{mnt, errAnonymousVolumeWithSubpath}
```

corhere reviewed (commented):
Almost there!

vvoland commented on internal/safepath/common.go:
```diff
@@ -0,0 +1,70 @@
+//
+// If isLocalTo(path, basepath) returns true, then
+//
+// 	filepath.Join(path, filepath.Rel(basepath, path))
+//
+// will always produce a path contained within basepath and
```
Yeah, sorry, this was meant to be `filepath.Join(basepath, filepath.Rel(basepath, path))`...
Anyway, yeah it's really the same thing repeated twice, so let's just leave:
```go
...
// If isLocalTo(path, basepath) returns true, then
//
//  filepath.Rel(basepath, path)
//
// will always produce an unrooted path with no `..` elements.
...
```

`filepath.Join(basepath, path)` doesn't really make sense in this case.

Both basepath and path are absolute paths, like:
`basepath=/var/lib/docker/volumes/volume/_data`
`path=/var/lib/docker/volumes/volume/_data/subpath`

vvoland commented on internal/safepath/common.go:
```diff
@@ -0,0 +1,66 @@
+}
+
+// isLocalTo reports whether path, using lexical analysis only, has all of these properties:
+//   - is within the subtree rooted at subpath
```
🤦🏻 Thanks, fixed this one! (...and the same mistake in the last paragraph)

vvoland commented:
Thanks! 🎉 
Converted to a draft for now - will merge after 25 is released. Unless we decide that we want to get this in for the 25.

vvoland commented:
Rebased, the only code change is moving the feature to 1.45 API version:

<details>
<summary>diff</summary>

```diff
diff --git a/api/server/router/container/container_routes.go b/api/server/router/container/container_routes.go
index cce8e4fbfe..dbb2b9beda 100644
--- a/api/server/router/container/container_routes.go
+++ b/api/server/router/container/container_routes.go
@@ -616,9 +616,6 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 					return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
 				}
 			}
-			if m.VolumeOptions != nil && m.VolumeOptions.Subpath != "" {
-				return errdefs.InvalidParameter(errors.New("VolumeOptions.Subpath needs API v1.44 or newer"))
-			}
 		}
 
 		// Creating a container connected to several networks is not supported until v1.44.
@@ -631,6 +628,14 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 		}
 	}
 
+	if versions.LessThan(version, "1.45") {
+		for _, m := range hostConfig.Mounts {
+			if m.VolumeOptions != nil && m.VolumeOptions.Subpath != "" {
+				return errdefs.InvalidParameter(errors.New("VolumeOptions.Subpath needs API v1.45 or newer"))
+			}
+		}
+	}
+
 	var warnings []string
 	if warn, err := handleMACAddressBC(config, hostConfig, networkingConfig, version); err != nil {
 		return err
diff --git a/integration/volume/mount_test.go b/integration/volume/mount_test.go
index 2d1ea0785e..bac8cc56db 100644
--- a/integration/volume/mount_test.go
+++ b/integration/volume/mount_test.go
@@ -21,7 +21,7 @@ import (
 )
 
 func TestRunMountVolumeSubdir(t *testing.T) {
-	skip.If(t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.44"), "skip test from new feature")
+	skip.If(t, versions.LessThan(testEnv.DaemonAPIVersion(), "1.45"), "skip test from new feature")
 
 	ctx := setupTest(t)
 	apiClient := testEnv.APIClient()
```

</summary>

thaJeztah reviewed (approved):
LGTM - let's get this one in 🎉

zmweske commented:
is the long syntax required or would it be possible to utilize short syntax? It currently seems to only work with long syntax but it would make sense to me that if a named volume exists with a subpath, it should work. If a named volume doesn't exist, it will look for folders in ./

See: https://docs.docker.com/compose/compose-file/05-services/#volumes

```yaml
services:
  example:
    image: example:latest
    volumes:
      - named-volume/subpath:/internal/dir:rw  # would utilize a named volume that exists in the compose file already
      - local-dir/subpath:/internal/dir:rw  # would use a local dir as no named volume exists
      - ./local-dir2/subpath:/internal/dir2:rw  # functionally the same as above
...
volumes:
  named-volume:
```

Thoughts?

neersighted commented:
This is the wrong issue to discuss Compose's implementation. That being said, we're deliberately avoiding short/ambiguous formats (e.g. the implicit behavior you describe) for new features like this, for (hopefully obvious) reasons of explicit being better than implicit.

Less typing sounds nice, until it leads to a bevy of issue reports, wasted hours debugging, or (in the worst case) security issues.

Please open a discussion on https://github.com/compose-spec/compose-spec for questions, or open an issue if you have a concrete proposal/change to suggest.

thaJeztah commented:
Yeah, I don't think we want to overload the short-form syntax more. Also not sure if such a format should be defined by the compose-spec first (as there's an expectation for the short-form format to be the same as is used for `-v` short-form).
